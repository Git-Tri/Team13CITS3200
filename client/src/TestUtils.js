const trueErrorLog = console.error;

/**
 * okay this is a weird one to explain
 * basically there an issue with our dev setup (web-pack/react/create react app)
 * where if you test when a component fails it will spam the error log with
 * a message telling you the component failed 
 * since we are testing that, this is completely unneccessary
 * as such i am filtering the console.error to exclude these messages 
 * to give a more readable test output
 * using monkey patching: https://davidwalsh.name/monkey-patching
 */
export function filterConsoleError()
{

    
    console.error = (e) => 
    {

        
        if(typeof(e) === "string" && ! e.includes("React will try to recreate this component tree from scratch using the error boundary you provided"))
        {

            trueErrorLog(e);
        
        }
        else if(typeof(e) !== "string")
        {

            trueErrorLog(e);

        }
        
    }

}

export function unfilterConsoleError()
{

    console.error = trueErrorLog;

}