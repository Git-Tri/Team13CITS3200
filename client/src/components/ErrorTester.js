import React from "react";

/**
 * Catches an error and uses the onError callback to allow tests to pass
 * this component is meant for testing purposes only
 * when testing place the error component as a child to this component 
 * then pass in an onError prop or check the state of this component
 * derived from: https://codepen.io/gaearon/pen/wqvxGa?editors=0010
 */
class ErrorTester extends React.Component {
    constructor(props) {
      super(props);
      this.state = { error: null, errorInfo: null };
    }
    
    componentDidCatch(error, errorInfo) {
      this.setState({
        error: error,
        errorInfo: errorInfo
      })

      this.props.onError(error,errorInfo);
    }
    
    render() {
      if (this.state.errorInfo) {
        return (
          <p> Good news an error occurred</p>
          );
      }
      return this.props.children;
    }  
  }

  export default ErrorTester;
