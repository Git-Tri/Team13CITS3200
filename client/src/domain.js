/**
 * A class representing unstructured data
 * for list you may want to leave the data field undefined
 * NOTE: this code is duplicated on client and server so is only tested on the server
 */
export class UnstructuredData
{

    constructor(id,matchid,title,author,url,published,extracted,data)
    {

        this.id = id;
        this.matchid = matchid;
        this.title = title;
        this.author = author;
        this.url = url;
        this.published = published;
        this.extracted = extracted;
        this.data = data; 

    }

}

/**
 * A class representing structured data, for lists you may want to leave data undefined
 */
export class StructuredData
{

    constructor(id,date,home,away,competitionID,competitionName,data)
    {

        this.id = id;
        this.date = date;
        this.home = home;
        this.away = away;
        this.competitionID = competitionID;
        this.competitionName = competitionName;
        this.data = data; 

    }

}

/**
 * a class representing a match 
 */
export class Match
{

    constructor(id,date,home,away,competitionID,competitionName)
    {

        this.id = id;
        this.date = date;
        this.home = home;
        this.away = away;
        this.competitionID = competitionID;
        this.competitionName = competitionName;

    }

}

class InsertMatch
{

    constructor(id,competitionID,date,home,away,data)
    {

        this.id = id;
        this.competitionID = competitionID;
        this.date = date;
        this.home = home;
        this.away = away;
        this.data = data;
          

    }

}

/**
 * A class representing an edit 
 */
export class Edit 
{

    constructor(editID,structuredDataID,unstructuredDataID,isCorpus,settings,replace,replaceWith,type,order)
    {

        this.editID = editID;
        this.structuredDataID = structuredDataID;
        this.unstructuredDataID = unstructuredDataID;
        this.isCorpus = isCorpus;
        this.settings = settings;
        this.replace = replace;
        this.replaceWith = replaceWith;
        this.type = type;
        this.order = order;
        
    }

}


export class Competition
{

    constructor(id,name,countryName,countryId)
    {

        this.id = id; 
        this.name = name; 
        this.countryName = countryName;
        this.countryId = countryId

    }

}

export class ImportRequest
{

    constructor(begin,end,compId)
    {

        this.begin = begin; //optional

        this.end = end; //optional

        this.compId = compId; //mandatory 

    }

}


export class SearchRequest
{

    constructor(type,value,field)
    {

        this.type = type;
        this.value = value;
        this.field = field; 

    }

}


export class User
{

    constructor(id, username, hash, admin, regkey, token, apikey)
    {

        this.id = id; 
        this.username = username; 
        this.hash = hash;
        this.admin = admin;
        this.regkey = regkey;
        this.token = token;
        this.apikey = apikey;

    }

}