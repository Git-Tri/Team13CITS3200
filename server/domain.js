/**
 * A class representing unstructured data
 * for list you may want to leave the data field undefined
 * NOTE: this code is duplicated on client and server so is only tested on the server
 */
class UnstructuredData
{

    constructor(id,matchid,title,author,url,published,extracted,data)
    {

        this.id = id;
        this.matchid = matchid;
        this.title = title;
        this.author = author;
        this.url = url;
        this.published = published;
        this.extracted = extracted;
        this.data = data; 

    }

}

/**
 * A class representing structured data, for lists you may want to leave data undefined
 */
class StructuredData
{

    constructor(id,date,home,away,competitionID,competitionName,data)
    {

        this.id = id;
        this.date = date;
        this.home = home;
        this.away = away;
        this.competitionID = competitionID;
        this.competitionName = competitionName;
        this.data = data; 

    }

}

/**
 * a class representing a match 
 */
class Match
{

    constructor(id,date,home,away,competitionID,competitionName)
    {

        this.id = id;
        this.date = date;
        this.home = home;
        this.away = away;
        this.competitionID = competitionID;
        this.competitionName = competitionName;

    }

}

class InsertMatch
{

    constructor(id,competitionID,date,home,away,data)
    {

        this.id = id;
        this.competitionID = competitionID;
        this.date = date;
        this.home = home;
        this.away = away;
        this.data = data;
          

    }

}

/**
 * A class representing an edit 
 */
class Edit 
{

    constructor(editID,structuredDataID,unstructuredDataID,isCorpus,settings,replace,replaceWith,type)
    {

        this.editID = editID;
        this.structuredDataID = structuredDataID;
        this.unstructuredDataID = unstructuredDataID;
        this.isCorpus = isCorpus;
        this.settings = settings;
        this.replace = replace;
        this.replaceWith = replaceWith;
        this.type = type;
        
    }

}

class Competition
{

    constructor(id,name,countryName,countryId)
    {

        this.id = id; 
        this.name = name; 
        this.countryName = countryName;
        this.countryId = countryId

    }

}

class User
{

    constructor(id, username, hash, admin, regkey)
    {

        this.id = id; 
        this.username = username; 
        this.hash = hash;
        this.admin = admin;
        this.regkey = regkey;

    }

}


class ImportRequest
{

    constructor(begin,end,compId)
    {

        this.begin = begin; //optional

        this.end = end; //optional

        if(typeof(compId !== "string"))
        {

            throw new Error("Competition must exist and of type string");

        }

        this.compId = compId; //mandatory 

    }

}

class SearchRequest
{

    constructor(type,value,field)
    {

        this.type = type;
        this.value = value;
        this.field = field; 

    }

}

//export all classes 
module.exports = 
{

    UnstructuredData,StructuredData,Match,Edit,Competition,ImportRequest, InsertMatch, SearchRequest, User

}