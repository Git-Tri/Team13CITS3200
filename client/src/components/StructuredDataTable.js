import React, { Component } from 'react';
import { Table } from 'semantic-ui-react';
import { StructuredData} from "../domain";

/**
 * shows a table of structured data 
 * has the ability to select a row 
 */
class StructuredDataTable extends Component
{

    /**
     * constructs a new structured data table
     * data is mandatory however onSelect is optional 
     * @param {*} props the props passed in
     */
    constructor(props)
    {

        super(props);

        let selectFunc;

        if(this.props.onSelect != undefined)
        {
            selectFunc = (data) => 
                {
                    
                    this.setState({activeRow: data.id})

                    this.props.onSelect(data);

                }
            

        }
        else
        {

            selectFunc = () => {};

        }

        this.state = {activeRow: -1,selectFunc: selectFunc};

    }

    /**
     * renders the function
     * also checks if the data is valid 
     * only checks the fields which are relevant to this component 
     */
    render()
    {

        if(! Array.isArray( this.props.items))
        {

            throw Error("Props.items should contain a list of structured data")

        }
        if(this.props.items.every((i) => i != undefined && i != null) == false)
        {

            throw new Error("props.data must be an instance of structured data")

        }
        //borken into each predicate for easier debugging
        if(this.props.items.every((i) => 
        {
            let isIdValid = typeof(i.id) === "number"
            let isDateValid = i.date instanceof Date;
            let isHomeValid = typeof(i.home) === "string";
            let isAwayValid = typeof(i.away) === "string";
            let isCompNameValid = typeof(i.competitionName) === "string";
            let isPlanValid = typeof(i.plan) === "string"
            return isIdValid && isDateValid && isHomeValid && isAwayValid && isCompNameValid && isPlanValid;
        }) === false)
        {

            throw new Error("every piece of data should have valid date, home, away, comp name and plan");

        }

        let isSelectable = this.props.onSelect !== undefined;



        let rows = this.props.items.map((item, index) => (
            <StructuredDataRow 
                key={item.id} 
                data={item} 
                onSelect={this.state.selectFunc} 
                isActive={item.id == this.state.activeRow}
            />
        ))

        return(
        <Table striped selectable={isSelectable} >        
            <Table.Header>
                <Table.Row>
                    <Table.HeaderCell>Date</Table.HeaderCell>
                    <Table.HeaderCell>Match</Table.HeaderCell>
                    <Table.HeaderCell>Competition</Table.HeaderCell>
                    <Table.HeaderCell>Plan</Table.HeaderCell>
                </Table.Row>
            </Table.Header>  
            <Table.Body>
                {rows}
            </Table.Body>        
        </Table>
        );
    }
}

/**
 * Represents a row of the structured data table
 * as this component isn't exported no additional validation is done 
 * by this component 
 * @param {*} props 
 */
function StructuredDataRow(props)
{

    if((props.data instanceof StructuredData) == false)
    {

        throw Error("props.data must be an instance of structured data");

    }
    
    let date = props.data.date;

    let dateString = date.getDay() + "/" + date.getMonth() + "/" + date.getFullYear();

    let versus = props.data.home + " vs " + props.data.away;

    let selectFunc = props.onSelect != undefined ? props.onSelect : () => {};

    return(
        <Table.Row onClick={() => selectFunc(props.data)} active={props.isActive}>
            <Table.Cell>{dateString}</Table.Cell>
            <Table.Cell>{versus}</Table.Cell>
            <Table.Cell>{props.data.competitionName}</Table.Cell>
            <Table.Cell>{props.data.plan}</Table.Cell>
        </Table.Row>
    );
    
}

export default StructuredDataTable;