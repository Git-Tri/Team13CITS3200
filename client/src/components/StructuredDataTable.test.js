import StructuredDataTable from "./StructuredDataTable";
import renderer from 'react-test-renderer';
import React from 'react';
import ErrorTester from "./ErrorTester";
import {StructuredData} from "../domain";
import {filterConsoleError,unfilterConsoleError} from "../TestUtils";
import {shallow, mount} from "enzyme";

import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });


describe("Structured Data Table Tests",function()
{



    beforeAll(() =>
        {

            return new Promise(function(resolve,reject)
            {

                filterConsoleError();

                resolve();

            })

        });

    
    const testStructuredData = [
        new StructuredData(1,new Date("1991-04-20T00:00:00.000Z"),"some team","some other team",1,"some comp",{}),
        new StructuredData(2,new Date("1991-04-20T00:00:00.000Z"),"some team","some other team",1,"some comp",{})
        ];
    

    test("Should throw error when given undefined items",(done) => 
    {

    renderer.create(<ErrorTester              
             onError={(e,em) => {
                    expect(e.message).toEqual("Props.items should contain a list of structured data"); 
                    done();}
                    }>
                 <StructuredDataTable ></StructuredDataTable>
             </ErrorTester>
            );       

    });

    test("Should throw error when given non-array items",(done) => 
    {

        renderer.create(<ErrorTester              
            onError={(e,em) => {
                   expect(e.message).toEqual("Props.items should contain a list of structured data"); 
                   done();}
                   }>
                <StructuredDataTable items={"not an item"}></StructuredDataTable>
            </ErrorTester>
           );       

    });

    test("Should throw error when one item is undefined",(done) => 
    {

        let missingItemData = testStructuredData.slice();

        missingItemData.push(undefined);

        renderer.create(<ErrorTester              
            onError={(e,em) => {
                   expect(e.message).toEqual("props.data must be an instance of structured data"); 
                   done();}
                   }>
                <StructuredDataTable items={missingItemData}></StructuredDataTable>
            </ErrorTester>
           );       

    });

    test("Should throw error when item id wrong type",(done) => 
    {

        let wrongIDItemData = testStructuredData.slice();

        wrongIDItemData.push(new StructuredData("not an id",new Date(),"some team","some other team",1,"some comp",{}));

        renderer.create(<ErrorTester              
            onError={(e,em) => {
                   expect(e.message).toEqual("every piece of data should have valid date, home, away, comp name and plan"); 
                   done();}
                   }>
                <StructuredDataTable items={wrongIDItemData}></StructuredDataTable>
            </ErrorTester>
           );       

    });


    test("Should render correctly for correct input",() => 
    {        
                 
        let Component = renderer.create(<StructuredDataTable items={testStructuredData.slice()}/>)

        expect(Component.toJSON()).toMatchSnapshot();

    })

    test("Should render correctly for correct input with onSelect handler",() => 
    {        
        
        let Component = renderer.create(<StructuredDataTable onSelect={() => {}} items={testStructuredData.slice()}/>)

        expect(Component.toJSON()).toMatchSnapshot();

    })

    test("Should call callback when onselect is trigerred",(done) => 
    {

        let wrapper = mount(<StructuredDataTable onSelect={() => done()} items={testStructuredData.slice()}/>)

        wrapper.state().selectFunc(testStructuredData[0]);


    })

    test("Should render correctly with selected row",() => 
    {

        let wrapper = shallow(<StructuredDataTable onSelect={() => {}} items={testStructuredData.slice()}/>)

        wrapper.state().selectFunc(testStructuredData[0]);

        expect(wrapper).toMatchSnapshot();


    })

    test("Should render correctly for correct with no data",() => 
    {        
        
        let Component = renderer.create(<StructuredDataTable onSelect={() => {}} items={[]}/>)

        expect(Component.toJSON()).toMatchSnapshot();

    })

    afterAll(() =>
    {

        return new Promise(function(resolve,reject)
        {

            unfilterConsoleError();

            resolve();

        })

    });

})

