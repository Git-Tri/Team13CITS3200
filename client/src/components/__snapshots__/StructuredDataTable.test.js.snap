// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Structured Data Table Tests Should render correctly for correct input 1`] = `
<table
  className="ui striped table"
>
  <thead
    className=""
  >
    <tr
      className=""
    >
      <th
        className=""
      >
        Date
      </th>
      <th
        className=""
      >
        Match
      </th>
      <th
        className=""
      >
        Competition
      </th>
      <th
        className=""
      >
        Plan
      </th>
    </tr>
  </thead>
  <tbody
    className=""
  >
    <tr
      className=""
      onClick={[Function]}
    >
      <td
        className=""
      >
        6/3/1991
      </td>
      <td
        className=""
      >
        some team vs some other team
      </td>
      <td
        className=""
      >
        some comp
      </td>
    </tr>
    <tr
      className=""
      onClick={[Function]}
    >
      <td
        className=""
      >
        6/3/1991
      </td>
      <td
        className=""
      >
        some team vs some other team
      </td>
      <td
        className=""
      >
        some comp
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Structured Data Table Tests Should render correctly for correct input with onSelect handler 1`] = `
<table
  className="ui selectable striped table"
>
  <thead
    className=""
  >
    <tr
      className=""
    >
      <th
        className=""
      >
        Date
      </th>
      <th
        className=""
      >
        Match
      </th>
      <th
        className=""
      >
        Competition
      </th>
      <th
        className=""
      >
        Plan
      </th>
    </tr>
  </thead>
  <tbody
    className=""
  >
    <tr
      className=""
      onClick={[Function]}
    >
      <td
        className=""
      >
        6/3/1991
      </td>
      <td
        className=""
      >
        some team vs some other team
      </td>
      <td
        className=""
      >
        some comp
      </td>
    </tr>
    <tr
      className=""
      onClick={[Function]}
    >
      <td
        className=""
      >
        6/3/1991
      </td>
      <td
        className=""
      >
        some team vs some other team
      </td>
      <td
        className=""
      >
        some comp
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Structured Data Table Tests Should render correctly for correct with no data 1`] = `
<table
  className="ui selectable striped table"
>
  <thead
    className=""
  >
    <tr
      className=""
    >
      <th
        className=""
      >
        Date
      </th>
      <th
        className=""
      >
        Match
      </th>
      <th
        className=""
      >
        Competition
      </th>
      <th
        className=""
      >
        Plan
      </th>
    </tr>
  </thead>
  <tbody
    className=""
  />
</table>
`;

exports[`Structured Data Table Tests Should render correctly with selected row 1`] = `
<Table
  as="table"
  selectable={true}
  striped={true}
>
  <TableHeader
    as="thead"
  >
    <TableRow
      as="tr"
      cellAs="td"
    >
      <TableHeaderCell
        as="th"
      >
        Date
      </TableHeaderCell>
      <TableHeaderCell
        as="th"
      >
        Match
      </TableHeaderCell>
      <TableHeaderCell
        as="th"
      >
        Competition
      </TableHeaderCell>
      <TableHeaderCell
        as="th"
      >
        Plan
      </TableHeaderCell>
    </TableRow>
  </TableHeader>
  <TableBody
    as="tbody"
  >
    <StructuredDataRow
      data={
        StructuredData {
          "away": "some other team",
          "competitionID": 1,
          "competitionName": "some comp",
          "data": Object {},
          "date": 1991-04-20T00:00:00.000Z,
          "home": "some team",
          "id": 1,
        }
      }
      isActive={true}
      key="1"
      onSelect={[Function]}
    />
    <StructuredDataRow
      data={
        StructuredData {
          "away": "some other team",
          "competitionID": 1,
          "competitionName": "some comp",
          "data": Object {},
          "date": 1991-04-20T00:00:00.000Z,
          "home": "some team",
          "id": 2,
        }
      }
      isActive={false}
      key="2"
      onSelect={[Function]}
    />
  </TableBody>
</Table>
`;
